### DISCUSSION TITLE
想讨论下 IDE 分布式弹性架构的可能性

### DISCUSSION THREAD

**Discussion by geekeren** [opensumi/core#2665] [Ideas]
由于 IDE 后端使用独立容器运行，同时需要较高的资源保障。是否可以考虑将后端服务进行一定的进程拆分，使用容器弹性伸缩的方式来动态扩展计算资源。
当用户处于非活动状态时，可以保障个基本运行资源（比如 1核2G），比如调试、编译 等高资源资源的操作可以动态启动新的 docker 容器节点（比如 8核16G）去运行。

顺便提一点，用户浏览器端的 WebContainer 也可以作为弹性计算的节点类型之一，最好能和后端节点注册扩展方式保持同样的技术架构。

整体来讲，就是 opensumi 是一个可以动态扩展运行节点的分布式架构，根据运行时所需的资源大小进行弹性伸缩运行节点；节点形式包括 服务器的 docker / k8s 等容器节点 和 用户端浏览器的 web container

**Comment by erha19** [comment]
@geekeren 目前在 OpenSumi 中，针对后端进程是有一定的拆分的（如插件进程、调试进程、终端进程），对于 WebIDE 来说，前端的核心交互是与 WebSocket 的通信达成的，可以考虑独立部署前端资源，而将后端服务独立进行部署，进一步可将插件进程...等做部署拆分及弹性，这块目前我们内部已经应用上了，方案上是可行。

**Reply by geekeren** [reply]
目前前端我们也是单独部署的，后端进程都是在一个容器里。插件进程的弹性部署 得和 k8s 等容器平台结合吧，这一块现在就有方案了吗？
其次，进程分离是第一步，比较麻烦的是 后端的多个进程怎么共享磁盘吧，在一台机器上还好，多个容器共享同一磁盘应该比较麻烦。

**Comment by Ricbet** [comment]
@geekeren 多容器架构我们有对外做一些输出分享的，你可以看一下
QCon：https://qcon.infoq.cn/2022/shanghai/presentation/5136, Cloud IDE 技术架构的多容器章节
前端早早聊：https://www.zaozao.run/video/c34/c34-2, 云原生 Cloud IDE 的双容器技术章节（可能需要收费）

**Reply by geekeren** [reply]
这个我理解的是 分离 IDE 服务进程 和 用户的研发进程在不同容器里，是不是还是有常驻的两个容器在跑。能不能做到动态去起容器，甚至不止是双容器。比如，有一个场景，用户在平时写代码时，我们使用两个容器在跑服务进程和研发进程，当用户进行高耗资源"构建” 动作时，可以动态启动一个 构建进程跑在独立的容器，执行完成后容器就退出，如果这一部分跑在研发容器里也有可能把研发进程 CPU 打满 被容器服务给kill掉。

**Comment by geekeren** [comment]
@Ricbet 多容器 opensumi 这边有标准化的部署方案吗

**Reply by geekeren** [reply]
@erha19 有把插件进程跑在独立容器的具体实现代码吗

**Reply by erha19** [reply]
没有具体代码  ... 据我了解是通过 `AppConfig.extensionConnectOption` 配置去访问的插件进程地址，而插件进程的独立部署可以是起一个同样的服务或他们更合适的方法了

**Reply by geekeren** [reply]
@erha19 @Ricbet 怎么把插件进程独立部署出去，不是很清楚。是在另外一个 容器里面用 node 跑 ext-host.js 吗？具体启动参数和主进程的配置不是很明白

**Reply by geekeren** [reply]
@erha19 插件进程独立部署，需要完整跑一个相同的 node 服务器？还是可以单独只跑 extServer

**Reply by Ricbet** [reply]
@geekeren 
多容器目前还没有标准化的部署方案，内部也只是做了双容器架构的部署。不过可以给你提供一些思路。

首先双容器下需要解决两个核心问题
1. 插件进程如何单独运行
2. 容器间如何通信

为了能够保证不同容器之间能够正常通信，可以将原本插件进程与后端服务的 IPC 通信改成 RPC 的方式，那么就需要额外一个常驻的插件管理进程，来做插件进程的创建和销毁，目前 opensumi 里已经做了这部分工作，相关代码在:[ExtensionHostProxyManager](https://github.com/opensumi/core/blob/main/packages/extension/src/node/extension.host.proxy.manager.ts#L20)
使用方式是作为 IExtensionHostManager 这个 token 的 di 注入进去
```typescript
injector.addProviders({
  token: IExtensionHostManager,
  useFactory(injector) {
    return injector.get(ExtensionHostProxyManager, [{
      port: app.config.extHost.port,
    }]);
  },
});
```

然后插件进程的单独运行有个 [ext.host.proxy](https://github.com/opensumi/core/blob/main/packages/extension/src/hosted/ext.host.proxy.ts) 文件，直接 node 运行就行。

因为不同容器之间是同一个网络，可以仍旧采用基于 socket 文件的 ipc 方式，可以将 ext.host.proxy.js 文件以及 ipc socket 文件放在共享目录里，保证不同容器能够正常读取
```typescript
// appconfig
{
// 插件进程文件需要放在共享目录
extHost: 'xxx',
// 插件 socket 通信目录需要放置在共享目录
extHostIPCSockPath: 'xxx'
}
```

更多的细节可以阅读 extension/src/hosted 这个目录下的代码

**Reply by geekeren** [reply]
Sock 文件和 TCP port的方式是不是二选一。sock的文件路径是不确定的、每次随机生成的吧

**Reply by geekeren** [reply]
app.config.extHost.port 的端口配置和 ext.host.proxy启动端口是怎么保持一致的呢， ext.host.proxy 我没看到设置端口

**Reply by Ricbet** [reply]
sock 文件的生成默认都是在 tmp/sumi-ipc 目录下的 https://github.com/opensumi/core/blob/main/packages/core-common/src/utils/ipc.ts#L19

**Reply by Ricbet** [reply]
示例代码里的 `port: app.config.extHost.port` 这一段就是端口设置了

**Reply by geekeren** [reply]
ext.host.proxy 没有配置端口的地方

**Reply by geekeren** [reply]
@Ricbet 插件进程和主进程 需要有哪些目录是共享的，可以给个清单吗

**Comment by hacke2** [comment]
ext.host.proxy 可以自定义，不需要用框架默认提供的：

```js
import { ExtHostProxy } from '@opensumi/ide-extension/lib/hosted/ext.host.proxy-base';

const extHostProxy = new ExtHostProxy({
  socketConnectOpts: {
    port: 12345,
  },
});

extHostProxy.init();
```

之后可以在另外一个容器里拉起这个代理进程。

共享目录存在的意义就是可以在 ideserver 进程启动的时候将插件进程和插件进程代理进程复制到一个地方可以让主研发容器读到，目前就是插件进程和插件进程代理进程。

PS：双容器不只是插件进程需要改造，终端也需要改造，因为 ideserver 默认终端读取的是 ideserver 容器的终端，这块改造逻辑和插件进程改造逻辑类似，在主研发容器有一个 pty 的代理进程，这块在@opensumi/ide-terminal-next也有类似的视线

**Reply by geekeren** [reply]
比如用户代码目录、插件目录、配置目录 需要设置为双容器共享吗。其次软件环境，比如支持golang语言，插件进程的容器是否需要go的环境。还是插件的操作大部分被代理到主进程，插件进程的容器只需要一个 Nodejs环境就行

**Reply by hacke2** [reply]
1. 用户代码目录、插件目录、配置目录都应该是贡献目录，因为两个容器都能访问到，蚂蚁内部的实践是共享 /home 目录
2. 插件进程的容器就是用户研发容器，如果需要你要安装的语言服务插件依赖了对应语言软件，那么用户研发容器也要对应语言软件
3. 启动插件进程代理进程和插件进程是需要有 Node 环境，所以基础镜像需要有 Node
